core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -609,6 +617,27 @@ private class BufferReleasingInputStream(
   override def reset(): Unit = delegate.reset()
 }
 
+/**
+ * A listener to be called at the completion of the ShuffleBlockFetcherIterator
+ * @param data the ShuffleBlockFetcherIterator to process
+ */
+private class ShuffleFetchCompletionListener(var data: ShuffleBlockFetcherIterator)
+  extends TaskCompletionListener {
+
+  override def onTaskCompletion(context: TaskContext): Unit = {
+    if (data != null) {
+      data.cleanup()
+      // Null out the referent here to make sure we don't keep a reference to this
+      // ShuffleBlockFetcherIterator, after we're done reading from it, to let it be
+      // collected during GC. Otherwise we can metadata on block locations(blocksByAddress)
```

Comment:
What do you mean by `metadata on block locations`?

Reply:
@advancedxy the blocksByAddress variable noted in the brackets.

Reply:
> Otherwise we can metadata on block locations(blocksByAddress)

do you mean `we can hold metadata ...`

Reply:
@cloud-fan Yes, it's fixed! thank you!